Microsoft.Toolkit.Uwp.UI.Controls/FlexPanel/FlexPanel.cs

```diff
@@ -0,0 +1,1613 @@
+                            Frame_size2_i = 2;
+                            break;
+                    }
+
+                    _ordered_indices = null;
+                    if (item.ShouldOrderChildren && item.Count > 0)
+                    {
+                        var indices = new int[item.Count];
+
+                        // Creating a list of item indices sorted using the children's `order'
+                        // attribute values. We are using a simple insertion sort as we need
+                        // stability (insertion order must be preserved) and cross-platform
+                        // support. We should eventually switch to merge sort (or something
+                        // else) if the number of items becomes significant enough.
+                        for (int i = 0; i < item.Count; i++)
+                        {
+                            indices[i] = i;
+                            for (int j = i; j > 0; j--)
+                            {
+                                int prev = indices[j - 1];
+                                int curr = indices[j];
+                                if (item.Children[prev].Order <= item.Children[curr].Order)
+                                {
+                                    break;
+                                }
+
+                                indices[j - 1] = curr;
+                                indices[j] = prev;
+                            }
+                        }
+
+                        _ordered_indices = indices;
+                    }
+
+                    Flex_dim = 0;
+                    Flex_grows = 0;
+                    Flex_shrinks = 0;
+
+                    Reverse2 = false;
+                    Wrap = item.Wrap != FlexWrap.NoWrap;
+                    if (Wrap)
+                    {
+                        if (item.Wrap == FlexWrap.Reverse)
+                        {
+                            Reverse2 = true;
+                            Pos2 = Align_dim;
+                        }
+                    }
+                    else
+                    {
+                        Pos2 = Vertical ? item.PaddingLeft : item.PaddingTop;
+                    }
+
+                    Need_lines = Wrap && item.AlignContent != FlexAlignContent.Start;
+                    Lines = null;
+                    Lines_sizes = 0;
+                }
+
+                public FlexItem Child_at(FlexItem item, int i) =>
+                    item.Children[_ordered_indices?[i] ?? i];
+
+                public void Cleanup()
+                {
+                    _ordered_indices = null;
+                    Lines = null;
+                }
+            }
+
+            IEnumerator IEnumerable.GetEnumerator() =>
+                ((IEnumerable<FlexItem>)this).GetEnumerator();
+
+            IEnumerator<FlexItem> IEnumerable<FlexItem>.GetEnumerator() =>
+                (Children ?? System.Linq.Enumerable.Empty<FlexItem>()).GetEnumerator();
```

Comment:
Nit (potentially overkill, but easy to do): we could remove these two methods and add a pattern-based `GetEnumerator` returning the `List<T>.Enumerator` type, to get better performance and avoid the enumerator allocation. As in:

```csharp
private static readonly List<FlexItem> Empty = new List<FlexItem>();

public List<FlexItem>.Enumerator GetEnumerator() => (Children ?? Empty).GetEnumerator();
```

Similarly, the `: IEnumerable<FlexItem>` declaration for the class could be dropped too.
Code using the previous `GetEnumerator` APIs wouldn't need changes, as the C# compiler would automatically see and use this new version.

Reply:
Very helpful.

Would it be ok if I were to do something slightly different?  Applying the suggested change, verbatim, breaks the only use outside of the `FlexItem` class, at line 621 of FlexPanel.cs:

``` csharp
            var deleteCandidates = _root.GetEnumerator().ToList();
```

So, rather than:
``` csharp
           public List<FlexItem>.Enumerator GetEnumerator() => (Children ?? Empty).GetEnumerator();
```

how about:
```csharp
           internal List<FlexItem> LogicalChildren => Children is null
                ? new List<FlexItem>(Children)
                : Empty;
```

which in turn would change line 621 to:
``` csharp
            var deleteCandidates = _root.LogicialChildren;
```

Again, very interested in your insights.

Reply:
Yup, that breaks because `ToList` is a LINQ extension working on `IEnumerable<T>`. That was just the basic idea, some other tweaks would be needed to make that work properly, as you've noticed 🙂

I don't like that `LogicalChildren` property for two reasons:

- It does work (specifically, create a new list and copy items), so it should be exposed as a method to indicate that it's not cheap
- It returns an empty list that can be mutated accidentally, essentially breaking it in this scenario

I would suggest doing:

- The proposed optimized `GetEnumerator` method to power the two `foreach` loops that use it
- To have some other method (eg. `CopyChildrenAsList`) that will either return `new List(Children)`, or just a new list

This way we'll both get fast enumeration in those two `foreach` loops, and a safe and faster `ToList` version 👍

Reply:
Very helpful.  Thank you.